internal/tools/aws/database_multitenant.go

```diff
@@ -0,0 +1,637 @@
+	}
+
+	for _, databaseCluster := range databaseClusters {
+		dbInstallations, err := databaseCluster.GetInstallations()
+		if err != nil {
+			return errors.Wrapf(err, "could not get installations from database cluster ID %s", databaseCluster.ID)
+		}
+
+		if dbInstallations.Contains(d.installationID) {
+			unlocked, err := d.acquireLock(databaseCluster.ID, store)
+			if err != nil {
+				return errors.Wrapf(err, "could not lock database cluster ID %s", databaseCluster.ID)
+			}
+			defer unlocked(logger)
+
+			if !dbInstallations.Remove(d.installationID) {
+				return errors.Errorf("could not remove installation id %s from database cluster installations", d.installationID)
+			}
+
+			err = databaseCluster.SetInstallations(dbInstallations)
+			if err != nil {
+				return errors.Wrapf(err, "could not set installations in database cluster %s", databaseCluster.ID)
+			}
+
+			err = store.UpdateDatabaseCluster(databaseCluster)
+			if err != nil {
+				return errors.Wrapf(err, "could not set update database cluster %s", databaseCluster.ID)
+			}
+
+			rdsCluster, err := d.describeRDSCluster(databaseCluster.ID)
+			if err != nil {
+				return errors.Wrap(err, "could not describe RDS cluster")
+			}
+
+			err = d.updateCounterTag(rdsCluster.DBClusterArn, dbInstallations.Size())
+			if err != nil {
+				return errors.Wrapf(err, "failed to update counter tag in RDS cluster id %s", databaseCluster.ID)
+			}
+
+			return nil
+		}
+	}
+
+	databaseSecretName := RDSMultitenantSecretName(d.installationID)
+
+	_, err = d.client.Service().secretsManager.DeleteSecret(&secretsmanager.DeleteSecretInput{
+		SecretId: aws.String(databaseSecretName),
+	})
+	if err != nil && !IsErrorCode(err, secretsmanager.ErrCodeResourceNotFoundException) {
+		return errors.Wrapf(err, "failed to delete RDS database secret name %s", databaseSecretName)
+	}
+
+	logger.Info("RDS database installation teardown complete")
+
+	return nil
+}
+
+// Snapshot creates a snapshot of single database.
+func (d *RDSMultitenantDatabase) Snapshot(store model.InstallationDatabaseStoreInterface, logger log.FieldLogger) error {
+	return errors.New("not implemented")
+}
+
+// GenerateDatabaseSpecAndSecret creates the k8s database spec and secret for accessing a single database inside
+// a multitenant RDS cluster.
+func (d *RDSMultitenantDatabase) GenerateDatabaseSpecAndSecret(store model.InstallationDatabaseStoreInterface, logger log.FieldLogger) (*mmv1alpha1.Database, *corev1.Secret, error) {
+	logger.Info("Setting up database spec and secret")
+	logger = logger.WithField("rds-multitenant-database", d.installationID)
+
+	databaseClusters, err := store.GetDatabaseClusters()
+	if err != nil {
+		return nil, nil, errors.Wrapf(err, "could not get installation id %", d.installationID)
+	}
+	if len(databaseClusters) < 1 {
+		return nil, nil, errors.Errorf("could not find installation id %", d.installationID)
+	}
+
+	var rdsCluster *rds.DBCluster
+	for _, databaseCluster := range databaseClusters {
```

Comment:
Either moving this for loop to a helper function or adding a comment saying what it does would aid readability here 0/5